### ISSUE TITLE
Running it along with Calico

### ISSUE THREAD

**Issue by mazzy89** [squat/kilo#5]
Is it possible to run it along with Calico? Anyone has tried it?

**Comment by squat** [comment]
Implementing Calico compatibility is possible (and used to be in the repo) however it was recently taken out because Calico’s default IPAM method was changed so that it no longer uses the podCIDR allocated by the K8s node controller.

To support this, Kilo will need to add a Calico client and read Calico IPPool CRs to determine the subnets allocated to each node.

If you’d like to give re-implementing it a shot, please let me know; I would be happy to review a PR. Otherwise, it’s high on my list for the project.

**Comment by mazzy89** [comment]
Thank you for replying. I'd be happy to give it a shot and see what I can do about that.

**Comment by outbackdingo** [comment]
any further progress on this?

**Comment by aboutte** [comment]
I am also interesting in this

**Comment by baurmatt** [comment]
We're also interested in in seeing this implemented! Our use case is to leverage the VPN feature to build site-to-site VPN for connecting legacy service with a Kubernetes cluster.

**Comment by unixfox** [comment]
> My understanding that this effort will block using Kilo on GKE, is that correct?

Yes it will, unfortunately kilo can't be deployed on GKE at the moment.

**Comment by KlavsKlavsen** [comment]
No progress on this? :(

**Comment by outbackdingo** [comment]
yeah can we get calico suport, alot of us have non-GKE environments

**Comment by abh1kg** [comment]
Was there any progress on this?

**Comment by aspekt112** [comment]
Running Kilo on top of a GKE cluster is still not possible?

**Comment by mazocode** [comment]
Can't we just re-add calico compatibility as it is in case the podCIDDR is the only issue? Calico IPools are great to have subnets per site and use different site specific or top of rack BGP peers.  The IP block assigned to a node is stored in a BlockAffinity CR. We are using a script to update the podCIDR from the generated blockaffinity after a new node joined so in our case this would always match. 

It is somewhat hacky but this is what we do to fix this issue:
```
nodeList=$(kubectl get nodes -o jsonpath='{.items[*].metadata.name}')

for node in $nodeList; do
    echo -n "* Node $node..."
    calicoName=$(kubectl get blockaffinities -o jsonpath='{.items[*].metadata.name}' | tr " " "\n" | awk "/$node/ {print $1}")
    echo -n " subnet=${calicoName}"
    calicoCIDR=$(kubectl get blockaffinities $calicoName -o jsonpath='{.spec.cidr}')
    echo -n ", ${calicoCIDR}"
    podCIDR=$(kubectl get node $node -o jsonpath='{.spec.podCIDR}')
    echo -n " == ${podCIDR}"
    if [ "${podCIDR}" == "${calicoCIDR}" ]; then
        echo " OK"
        continue
    fi
    if [ -z "${calicoCIDR}" ]; then
        echo " ERROR"
        echo "  Could not determine calico CIDR. Did you switch the nodes site? In this case make"
        echo "  sure there is only one blockaffinity object for this node."
        continue
    fi
    echo " ERROR"
    echo
    echo "!!!WARNING: Continuing might disrupt workloads running on the node!!!"
    echo
    echo -n "Shall we continue to fix it now (node should be drained before)? [y/n]? "
    read a
    if [ "$a" != "y" ]; then
        continue
    fi

    echo "Saving node yaml to ~/${node}.yaml..."
    kubectl get node $node -o yaml >~/${node}.yaml

    echo "Replacing CIDR (s/${podCIDR}/${calicoCIDR}/)..."
    sed -i "s/${podCIDR/\//\\/}/${calicoCIDR/\//\\/}/" ~/${node}.yaml
    echo "Deleting node..."
    kubectl delete node $node
    kubectl create -f ~/${node}.yaml

    echo "FIXED!"
done
```

**Comment by splitice** [comment]
This is a blocker for usage with LKE (linode) too.

**Comment by freexploit** [comment]
I made subnet being able to use podCIDR querying calico blockaffinity resource. That made calico start and stop giving this non-ready node error. Now I'm trying to understand the calico compatibility layer you wrote to se if I can make it work.

**Comment by LogeshGovi** [comment]
is there any progress?

**Comment by DodoLeDev** [comment]
I faced an unexpected behaviour by trying to deploy Kilo on Calico-networked clusters:

```
{"caller":"mesh.go:610","component":"kilo","error":"failed to add route \"dst: 10.4.0.1/32, via: 10.42.2.0, src: 1.2.3.4, dev: 0\": no such device","level":"error"}
```

It seems that traffic rules are applied on a specific device while not being a CNI. Which makes the [`cniDeviceIndex()`](https://github.com/squat/kilo/blob/495ad34d9c6156de963860cc0e78c4c8fd98680b/pkg/mesh/cni.go#L38) to return 0. Device number on which this rule is being applied.

I miss Calico support too :(

**Comment by DodoLeDev** [comment]
Turns out running Kilo with `--encapsulate=never` will prevent these errors from appearing.

However, this will only allow Kilo to encapsulate network across kubernetes clusters, and not across nodes (or subnets)

I will test it in the next week and will give you some news about that
